### PR TITLE
Handle init error events / better callback typing

### PR DESCRIPTION
**Pull request by qodesmith**
### Description & motivation

This PR adds the following:
* Better typing for the init callback function. As is done in the Node-style callbacks, any empty values are now explicitly set to `null`. The callback function now has 2 type overloads to better express this scenario. This will allow consumers to explicitly check for an error or an instance and handle their UI accordingly.
* The `CoinbasePixel` instance now has a new private `addErrorListener` method. This allows the pixel to provide the consumers callback function with errors from a variety of sources, such as a failure to parse app params.

### Testing & documentation

Added unit tests to cover the changes.

### CODE REVIEW

File: src/utils/CoinbasePixel.test.ts [hunk 1]
```diff
@@ -66,14 +66,24 @@ describe('CoinbasePixel', () => {
   it('should setup pixel ready listener', () => {
     createUntypedPixel(defaultArgs);
 
-    expect(onBroadcastedPostMessage).toHaveBeenCalledTimes(1);
+    expect(onBroadcastedPostMessage).toHaveBeenCalledTimes(2);
     expect(onBroadcastedPostMessage).toHaveBeenCalledWith('pixel_ready', {
       allowedOrigin: 'https://pay.coinbase.com',
       onMessage: expect.any(Function),
       shouldUnsubscribe: false,
     });
```
Comment (zack-winchell-cb) [inline src/utils/CoinbasePixel.test.ts:74]: nit: add `expect(onBroadcastedPostMessage).toHaveBeenCalledWith('error' /* etc */)`

File: src/utils/CoinbasePixel.test.ts [hunk 2]
```diff
@@ -66,14 +66,24 @@ describe('CoinbasePixel', () => {
   it('should setup pixel ready listener', () => {
     createUntypedPixel(defaultArgs);
 
-    expect(onBroadcastedPostMessage).toHaveBeenCalledTimes(1);
+    expect(onBroadcastedPostMessage).toHaveBeenCalledTimes(2);
```
Comment (zack-winchell-cb) [inline src/utils/CoinbasePixel.test.ts:69]: nit: This quantity check might want to move out to its own `it()` block

File: src/utils/CoinbasePixel.ts [hunk 3]
```diff
@@ -213,13 +216,28 @@ export class CoinbasePixel {
...
+        this.removeErrorListener?.();
         this.sendAppParams(() => {
           this.onReadyCallback?.();
         });
       },
     });
   };
 
+  private addErrorListener = (): void => {
+    this.removeErrorListener = this.onMessage('error', {
```
Comment (zack-winchell-cb) [inline src/utils/CoinbasePixel.ts:228]: Do we need to have removeErrorListener both `unsub()` here & this.unsubs.remove(unsub)`?

Comment (qodesmith) [reply src/utils/CoinbasePixel.ts:228]: `removeErrorListener` will unsub this event listener when initialization is _successful_. I'm thinking that the consumer's callback should only ever be called once, and that should only happen during initialization, _before_ calling the `.open` method.

Comment (zack-winchell-cb) [reply src/utils/CoinbasePixel.ts:228]: That all sounds right to me, but I'm thinking about the fact that `this.unsubs` will still have en entry for this unsubscriber after we've called it. It's definitely a smaller thing - I should've prefaced this comment with `nit:`

Comment (qodesmith) [reply src/utils/CoinbasePixel.ts:228]: Yea. `this.unsubs` will end up calling `window.removeEventListener(event, unsubFxn)`. That should be harmless if the listener was already removed.

**Review by lachiet — APPROVED** [review]

**Review by ignacio-palladino-cb — APPROVED** [review]

**Pull request merged**